File: mathsys/v6/python/ir.py
#^
#^  HEAD
#^

#> HEAD -> MODULES
from __future__ import annotations
from dataclasses import dataclass
from typing import Protocol, TypeVar, Generic
from zlib import compress


#^
#^  HELPERS
#^

#> HELPERS -> NODE
def node(node: Binary, binary: list[Binary], nodes: list[Binary]) -> Pointer:
    try: return Pointer(nodes.index(node))
    except ValueError: 
        nodes.append(node)
        binary.append(node)
        return Pointer(len(nodes) - 1)


#^
#^  PROTOCOL
#^

#> PROTOCOL -> DEFINITION
class BinaryEncodable(Protocol):
    def binary(self) -> Binary: ...

#> PROTOCOL -> TYPEVAR
Type = TypeVar("Type", bound = BinaryEncodable)


#^
#^  TYPES
#^

#> TYPES -> BINARY
@dataclass(kw_only = True, frozen = True)
class Binary:
    value: int
    width: int
    def __bytes__(self) -> bytes:
        value = Binary(
            value = self.value,
            width = self.width
        )
        while value.width % 8 != 0: value = value + Opcode(0x00)
        return compress(value.value.to_bytes(value.width // 8, byteorder = "little"), level = 9, wbits = -15)
    def __add__(self, other: Binary | BinaryEncodable) -> "Binary":
        if not isinstance(other, Binary): other = other.binary()
        return Binary(
            value = self.value | (other.value << self.width),
            width = self.width + other.width
        )

#> TYPES -> OPCODE
@dataclass(frozen = True)
class Opcode:
    value: int
    def binary(self) -> Binary:
        return Binary(
            value = self.value,
            width = 5
        )

#> TYPES -> POINTER
@dataclass(frozen = True)
class Pointer:
    value: int
    def binary(self) -> Binary: 
        return Binary(
            value = self.value, 
            width = 32
        )

#> TYPES -> SIGN
@dataclass(frozen = True)
class Sign:
    value: bool
    def binary(self) -> Binary: 
        return Binary(
            value = 1 if self.value else 0,
            width = 1
        )

#> TYPES -> OPTION
@dataclass(frozen = True)
class Option(Generic[Type]):
    value: Binary | BinaryEncodable | None
    def binary(self) -> Binary:
        if self.value is None:
            return Binary(value = 0, width = 1)
        else:
            return Binary(value = 1, width = 1) + self.value

#> TYPES -> BIGUINT
@dataclass(frozen = True)
class BigUint:
    value: int
    def binary(self) -> Binary:
        return Binary(
            value = self.value,
            width = 128
        )

#> TYPES -> STRING
@dataclass(frozen = True)
class String:
    value: str
    def binary(self) -> Binary:
        data = self.value.encode()
        length = len(data)
        return Binary(
            value = length,
            width = 16
        ) + Binary(
            value = int.from_bytes(data, byteorder = "little"),
            width = length * 8
        )

#> TYPES -> GROUP
@dataclass(frozen = True)
class Group:
    value: str
    def binary(self) -> Binary: return Binary(
        value = {
            "@Infinite": 1,
            "@Integer": 2,
            "@Natural": 3,
            "@Nexists": 4,
            "@Rational": 5,
            "@Tensor": 6,
            "@Undefined": 0,
            "@Variable": 7,
            "@Whole": 8
        }[self.value],
        width = 4
    )

#> TYPES -> VEC
@dataclass(frozen = True)
class Vec(Generic[Type]):
    values: list[Binary | BinaryEncodable]
    def binary(self) -> Binary:
        value = Binary(
            value = len(self.values),
            width = 32
        )
        for item in self.values: value += item
        return value

File: mathsys/dev/typescript/ir.ts
//^
//^ HEAD
//^

//> HEAD -> MODULES
import pako from "pako";


//^ 
//^ HELPERS
//^ 

//> HELPERS -> NODE
export function node(node: Binary, binary: Binary[], nodes: Binary[]): Pointer {
    const index = nodes.indexOf(node);
    if (index !== -1) {return new Pointer(index)}
    nodes.push(node);
    binary.push(node);
    return new Pointer(nodes.length - 1);
}


//^
//^ PROTOCOL
//^

//> PROTOCOL -> DEFINITION
interface BinaryEncodable {
    binary(): Binary;
}


//^
//^ TYPES
//^

//> TYPES -> BINARY
export class Binary implements BinaryEncodable {
    constructor(
        readonly value: bigint,
        readonly width: number
    ) {}
    toBytes(): Uint8Array {
        let value = new Binary(this.value, this.width);
        while (value.width % 8 !== 0) {value = value.add(new Opcode(0x00).binary())}
        const byteLength = value.width / 8;
        const bytes = new Uint8Array(byteLength);
        let val = value.value;
        for (let i = 0; i < byteLength; i++) {
            bytes[i] = Number(val & 0xffn);
            val >>= 8n;
        }
        return pako.deflateRaw(bytes, { level: 9 });
    }
    add(...others: (Binary | BinaryEncodable)[]): Binary {
        return others.reduce((accumulated: Binary, other: Binary | BinaryEncodable) => {
            const transformed = other instanceof Binary ? other : other.binary();
            return new Binary(
                accumulated.value | (transformed.value << BigInt(accumulated.width)),
                accumulated.width + transformed.width
            );
        }, this)
    }
    binary(): Binary {return this}
}

//> TYPES -> OPCODE
export class Opcode implements BinaryEncodable {
    constructor(
        readonly value: number
    ) {}
    binary(): Binary {return new Binary(BigInt(this.value), 5)}
}

//> TYPES -> POINTER
export class Pointer implements BinaryEncodable {
    constructor(
        readonly value: number
    ) {}
    binary(): Binary {return new Binary(BigInt(this.value), 32)}
}

//> TYPES -> SIGN
export class Sign implements BinaryEncodable {
    constructor(
        readonly value: boolean
    ) {}
    binary(): Binary {return new Binary(BigInt(this.value ? 1 : 0), 1)}
}

//> TYPES -> OPTION
export class Option<Type extends BinaryEncodable> implements BinaryEncodable {
    constructor(
        readonly value: Binary | BinaryEncodable | null
    ) {}
    binary(): Binary {return this.value === null ? new Binary(BigInt(0), 1) : new Binary(BigInt(1), 1).add(this.value)}
}

//> TYPES -> BIGUINT
export class BigUint implements BinaryEncodable {
    constructor(
        readonly value: bigint
    ) {}
    binary(): Binary {return new Binary(this.value, 128)}
}

//> TYPES -> STRING
export class String implements BinaryEncodable {
    constructor(
        readonly value: string
    ) {}
    binary(): Binary {
        const data = new TextEncoder().encode(this.value);
        const lengthBinary = new Binary(BigInt(data.length), 16);
        let valueBinary = new Binary(BigInt(0), data.length * 8);
        for (let i = 0; i < data.length; i++) {
            valueBinary = valueBinary.add(new Binary(BigInt(data[i]), 8));
        }
        return lengthBinary.add(valueBinary);
    }
}

//> TYPES -> GROUP
export class Group implements BinaryEncodable {
    constructor(
        readonly value: string
    ) {}
    binary(): Binary {return new Binary(
        BigInt({
            "@Infinite": 1,
            "@Integer": 2,
            "@Natural": 3,
            "@Nexists": 4,
            "@Rational": 5,
            "@Tensor": 6,
            "@Undefined": 0,
            "@Variable": 7,
            "@Whole": 8,
        }[this.value] as number),
        4
    )}
}

//> TYPES -> VEC
export class Vec<Type extends BinaryEncodable> implements BinaryEncodable {
    constructor(
        readonly values: (Binary | BinaryEncodable)[]
    ) {}
    binary(): Binary {
        let value = new Binary(BigInt(this.values.length), 32);
        for (const item of this.values) {
            value = value.add(item instanceof Binary ? item : item.binary());
        }
        return value;
    }
}